### ISSUE TITLE
Nginx pod not receiving traffic with Local traffic Policy, Layer 2 mode

### ISSUE THREAD

**Issue by asvillamar** [metallb/metallb#287]
**Is this a bug report or a feature request?**:
Bug/question

**What happened**:
I changed my nginx service's externalTrafficPolicy to Local and now my nginx pod no longer receives traffic

**What you expected to happen**:
The nginx pod will continue to get traffic, but with the source ip intact. Using Layer 2 mode

**Environment**:
- MetalLB version: 0.7.1
- Kubernetes version: latest
- BGP router type/version: NA
- OS (e.g. from /etc/os-release): centos7
- Kernel (e.g. `uname -a`): Linux K8SM1 3.10.0-862.3.2.el7.x86_64 #1 SMP Mon May 21 23:36:36 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

I have an nginx pod that listens for UDP on port 80, and redirects the UDP packet to 192.168.122.206:8080
I have a simple udp server that listens on 192.168.122.206:8080. This was working fine, but I needed to know the original source IP and port of the packet so I changed my service's traffic policy to local.
Now, the pod doesn't seem to get traffic.
I am running a single node bare metal cluster.
I have tried doing "kubectl logs pod-name" but nothing shows up, leading me to believe the pod isn't getting traffic at all.
I am making sure that my UDP packet is being sent to the external ip of the nginx service and port 80.

my nginx.conf from which I built the image:
```
user nginx;
worker_processes auto;
error_log /var/log/nginx/error.log;
pid /run/nginx.pid;
# Load dynamic modules. See /usr/share/nginx/README.dynamic.
include /usr/share/nginx/modules/*.conf;
events {
    worker_connections 1024;
}

stream {
server {
		listen 80 udp;
		proxy_pass 192.168.122.206:8080;
	}
}
```

My nginx deployment and service
```
apiVersion: apps/v1beta2
kind: Deployment
metadata:
  name: nginx
spec:
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: asvilla/custom_nginx2:first
        ports:
        - name: http
          containerPort: 80

---
apiVersion: v1
kind: Service
metadata:
  name: nginx
spec:
  externalTrafficPolicy: Local
  ports:
  - name: http
    port: 80
    protocol: UDP
    targetPort: 80
  selector:
    app: nginx
  type: LoadBalance
```

**Comment by asvillamar** [comment]
I have set verbosity of my pods and containers logs to 9. They show nothing new when I send the packet.
I also set verbosity to 9 for "kubectl describe service nginx"  and that doesn't show anything new when I send the packet. 
My best guess here is that something is going wrong with kube-proxy? Also the fact that my master is my only node might be affecting something, although when I set it up I untainted it and allowed the scheduler to treat it as a worker node.

**Comment by asvillamar** [comment]
Sorry I wish I could give you access but I can't, it's only a test cluster but my employers wouldn't allow it, appreciate it though. I am fairly new to this, how do I check if my kube-proxy is in ipvs mode? I haven't messed with it at all, I just followed the kubeadm instructions to get a basic single node cluster.

By the way 1. when I follow the layer 2 tutorial I have to disable and stop firewalld to get the nginx service an external ip, otherwise it is stuck at pending.

2. With the layer 2 tutorial, if I change the nginx service to local, I still get the default nginx page when I curl the external-ip:80 .  Don't know if this helps at all, thanks

**Comment by carlpett** [comment]
I'm trying to recreate this in a unit test, but not having any success. The layer2Controller.ShouldAnnounce and usableNodes functions seem to do what I understand they should do.
Is there some interaction with kube-proxy which might be to blame?

**Comment by danderson** [comment]
Ideally, what I need is ssh+root on a node, and k8s API access (i.e. a bearer token so I can auth to the apiserver). For the latter, I only need a token bound to the "view" ClusterRole, which gives me RO access to non-sensitive objects (i.e. I can't read secrets and such).

0 local endpoints sounds like a problem with the service or deployment specification. What does `kubectl get ep -oyaml -ningress-controller haproxy-ingress` say? Any viable backends?

This kind of thing is why I need to keep focusing on getting the e2e tests up and running :(

**Comment by carlpett** [comment]
```yaml
apiVersion: v1
kind: Endpoints
metadata:
  creationTimestamp: 2018-07-27T10:39:25Z
  labels:
    app: haproxy-ingress
    component: controller
  name: haproxy-ingress
  namespace: ingress-controller
  resourceVersion: "1115175"
  selfLink: /api/v1/namespaces/ingress-controller/endpoints/haproxy-ingress
  uid: 54961511-9189-11e8-a955-000c2907a29e
subsets:
- addresses:
  - ip: 10.244.2.113
    nodeName: agent-0
    targetRef:
      kind: Pod
      name: haproxy-ingress-6bff4db657-lwp5m
      namespace: ingress-controller
      resourceVersion: "1115164"
      uid: 46ee76f5-94a7-11e8-a906-000c2907a29e
  ports:
  - name: http
    port: 80
    protocol: TCP
  - name: https
    port: 443
    protocol: TCP
```

Everything works fine if I toggle the `Service` back to `externalTrafficPolicy=Cluster`, and stops working as soon as I set it to `Local`.

I'll try to get the remote access up tonight. It's getting late-isch, but let's see if I can get it done.

**Comment by carlpett** [comment]
@danderson I sent you an email with instructions for access, I used the email address your commits have as author. Correct?

**Comment by danderson** [comment]
Yes, that should be correct (just to verify: dave (at) natulte.net is my email).

I haven't received anything yet, FWIW.

**Comment by carlpett** [comment]
The tab had lost connectivity somehow, I didn't notice it was timing out. Now at least Gmail claims to be done with it :)

**Comment by danderson** [comment]
Yup, received. Thank you! I will only be able to take a look tonight (~4-6h from now in local time), but I'll report back with my findings (if any, hopefully :) ).

**Comment by carlosedp** [comment]
Hi @danderson  , I also have a K8s cluster running 1.10.6, MetalLB 0.7.1 and with Traefik 1.6.5 ingress controller. The Traefik service has the `externalTrafficPolicy: Local` and Traefik has logging enable but I still see the internal cluster IPs instead of the valid ones. I run on ARM64.

My topology is: ![this](https://cdn-images-1.medium.com/max/800/1*laQBZYKVPlcAxrv3Vuo65w.png)

MetalLb config:

```
apiVersion: v1
data:
  config: |
    address-pools:
    - name: default
      protocol: layer2
      addresses:
      - 192.168.1.16/28
kind: ConfigMap
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"config":"address-pools:\n- name: default\n  protocol: layer2\n  addresses:\n  - 192.168.1.16/28\n"},"kind":"ConfigMap","metadata":{"annotations":{},"name":"config","namespace":"metallb-system"}}
  creationTimestamp: 2018-03-08T21:50:31Z
  name: config
  namespace: metallb-system
  resourceVersion: "13690183"
  selfLink: /api/v1/namespaces/metallb-system/configmaps/config
  uid: b8e072a8-231a-11e8-8342-eea0ead58fde
```

Traefik Service:
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"labels":{"k8s-app":"traefik-ingress-lb"},"name":"traefik-ingress-service-external","namespace":"kube-system"},"spec":{"externalTrafficPolicy":"Local","loadBalancerIP":"192.168.1.21","ports":[{"name":"web","port":80,"protocol":"TCP"},{"name":"https","port":443,"protocol":"TCP"},{"name":"admin","port":8080,"protocol":"TCP"}],"selector":{"k8s-app":"traefik-ingress-lb-external"},"type":"LoadBalancer"}}
  creationTimestamp: 2018-05-05T01:51:09Z
  labels:
    k8s-app: traefik-ingress-lb
  name: traefik-ingress-service-external
  namespace: kube-system
  resourceVersion: "18308220"
  selfLink: /api/v1/namespaces/kube-system/services/traefik-ingress-service-external
  uid: c84daa63-5006-11e8-9474-eea0ead58fde
spec:
  clusterIP: 10.105.181.157
  externalTrafficPolicy: Local
  healthCheckNodePort: 32577
  loadBalancerIP: 192.168.1.21
  ports:
  - name: web
    nodePort: 31398
    port: 80
    protocol: TCP
    targetPort: 80
  - name: https
    nodePort: 31427
    port: 443
    protocol: TCP
    targetPort: 443
  - name: admin
    nodePort: 30525
    port: 8080
    protocol: TCP
    targetPort: 8080
  selector:
    k8s-app: traefik-ingress-lb-external
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - ip: 192.168.1.21
```

Some logs from traefik regargding access to the K8s dashboard that runs behind it:

```
10.40.0.1 - - [01/Aug/2018:20:14:01 +0000] "GET / HTTP/1.1" 404 19 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.103 Safari/537.36" 18 "backend not found" "/" 0ms
10.40.0.1 - - [01/Aug/2018:20:29:12 +0000] "GET / HTTP/1.1" 404 19 "-" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.103 Safari/537.36" 19 "backend not found" "/" 0ms
168.121.129.138 - - [01/Aug/2018:20:32:16 +0000] "GET / HTTP/1.1" 401 17 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.75 Safari/537.36 Google Favicon" 20 "Auth for dashboard.cloud.domain.com/" "/" 50ms
10.40.0.1 - - [01/Aug/2018:20:42:02 +0000] "GET / HTTP/1.1" 404 19 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36" 21 "backend not found" "/" 0ms
10.40.0.1 - - [01/Aug/2018:20:44:15 +0000] "GET / HTTP/1.1" 404 19 "-" "-" 22 "backend not found" "/" 0ms
10.40.0.1 - - [01/Aug/2018:21:03:45 +0000] "GET / HTTP/2.0" 200 527 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.75 Safari/537.36" 23 "dashboard.cloud.domain.com/" "http://10.38.0.14:9090" 8ms
10.40.0.1 - - [01/Aug/2018:21:03:45 +0000] "GET /api/appConfig.json HTTP/2.0" 200 64 "https://dashboard.cloud.domain.com/" "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.75 Safari/537.36" 24 "dashboard.cloud.domain.com/" "http://10.38.0.14:9090" 8ms
10.40.0.1 - - [01/Aug/2018:21:03:45 +0000] "GET /api/v1/login/status HTTP/2.0" 200 93 "https://dashboard.cloud.domain.com/" "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.75 Safari/537.36" 26 "dashboard.cloud.domain.com/" "http://10.38.0.14:9090" 10ms
```
If you want access to it let me know.

**Comment by danderson** [comment]
For the record, I said earlier that #291 might be the root cause. I don't think that's true any more. Switching traffic policies triggers a reconfiguration of the node routing by kube-proxy, but from MetalLB's perspective, the leader node should stay the same (it's not influenced by the traffic policy). So, there's no failover and no need to do any gratuitous ARP.

Still digging...

**Comment by jansmets** [comment]
I have the same problem. L2 set up. 
The external IP of my nodes and the LoadBalancer IP for the service are in the same subnet.

**Comment by Elegant996** [comment]
Experiencing same issue, near identical setup with VMs for nodes. Available for debugging as needed. Thanks!

**Comment by sonaikar** [comment]
Observed similar issue while exporting a service on NodePort.  Service works when "External Traffic Policy" is set to "Cluster" but stops working when it is set to "Local". One differentiating factor is my service is listening to UDP port instead of TCP. 

External Traffic Policy is working well in Kubernetes version "1.10.6" but having problems with "1.11.5". Anything that changed for policy or proxy in recent Kubernetes release?

**Comment by davidq2q** [comment]
Having same problem as well. Service stops working when External Traffic Policy is set to Local.

**Comment by schaze** [comment]
Just a thought as I was having some strange service problems myself. Can you try restarting the speaker pods in the daemonset after your service change and check if that helps? (Adding a label usually starts a new rollout)

**Comment by carlosedp** [comment]
After a long time battling this issue I found out that Weave Net was the culprit of my problem.

According to https://www.weave.works/docs/net/latest/kubernetes/kube-addon/#configuration-options, you need to set `NO_MASQ_LOCAL` to "1" in the deployment by using: `kubectl apply -f "https://cloud.weave.works/k8s/net?k8s-version=$(kubectl version | base64 | tr -d '\n')&env.NO_MASQ_LOCAL=1"`.

After this, I set my services to `externalTrafficPolicy: Local` and the source IP appeared.

**Comment by carlosedp** [comment]
After a long time battling this issue I found out that Weave Net was the culprit of my problem.

According to https://www.weave.works/docs/net/latest/kubernetes/kube-addon/#configuration-options, you need to set `NO_MASQ_LOCAL` to "1" in the deployment by using: `kubectl apply -f "https://cloud.weave.works/k8s/net?k8s-version=$(kubectl version | base64 | tr -d '\n')&env.NO_MASQ_LOCAL=1"`.

After this, I set my services to `externalTrafficPolicy: Local` and the source IP appeared.

**Comment by andrelop** [comment]
Hello,

I think I'm having the same issue here.
I'm using Kubernetes 1.15.0, MetalLN 0.7.3, flannel 0.11 in hosts-gw mode and kube-proxy in IPVS mode.

Also, Ingress Ngnix Service is configure as type LoadBalancer with externalTrafficPolicy=Local to preserve source IPs. It seems to work but suddenly after some random time traffic stops getting to Ingress Nginx to no app can be served.

If I change Ingress Nginx Service to externalTraffficPolicy: Cluster, everything works again, but then I cannot have source IP preserved.

I already tried using Calico, flannel and kube-router. All of them have the same issue. Is there a know workaround to this issue ?

**Comment by remche** [comment]
That's weird. Facing the same issue on only one of our clusters :

- kubernetes 1.13.5 (rke flavor)
- kube-proxy iptables mode
- canal (flannel+calico, rke default)
- MetalLB 0.7.3

Our test cluster has the same configuration, but do not face the issue...
Tried to restart speaker/controller as suggested by @schaze w/o success.

**Comment by remche** [comment]
Just spent some time on this issue... For my part, it wasn't a MetalLB issue but a kube-proxy misconfiguration.

You may be affected if you use an IP or a non resolvable address for `hostname-override` kube-proxy option. It has been already discussed [here](https://github.com/kubernetes/kubernetes/issues/57518), [here](https://github.com/kubernetes/kubernetes/issues/35641) and more...

To check, you can look at this line in kube-proxy log :
`W0919 09:11:08.173381   13772 proxier.go:314] invalid nodeIP, initializing kube-proxy with 127.0.0.1 as nodeIP`

In my case I was using IP for node address in the [RKE cluster.yml](https://rancher.com/docs/rke/latest/en/example-yamls/). Putting new nodes with the fdqn as address fixed it. Pretty sure I wont be the only one ;)

**Comment by Ufaowl** [comment]
I also stumbled upon this issue. I switch to Local mode and all traffic is gone. And I couldn't get any indication of why it's gone

**Comment by russellb** [comment]
I'm going to close this out as stale since it hasn't been updated in a long time.  Problems like this are *usually* not the fault of MetalLB, but whatever is implementing the data plane side.  I'd suggest a discussion on slack or the mailing list if someone runs into something similar to help narrow down what metallb is not doing before filing a follow-up issue.  Thanks!

**Comment by bit-project-de** [comment]
Hi all,

I got a similar Issue:

kubernetes: 1.22.8
Kubespray: 2.18.2
NginX Ingress Controller
calico stack
MetalLB 0.8.3


For me setting NginX Ingress Controller Service to Local Policy worked flawlessly until I did:

Starting from K8S 1.22.5 with Docker configuration I first migrated the cluster setup to Containerd. After Containerd Switch was done I updated Control Plane and workers to 1.22.8 which is the latest supported version for Kubespray Tag 2.18.2.

After I did this I did all the usual testing and validation of cluster components and everything was working as expected. But then after 1,5 hours or so all ingresses suddenly stopped working. 

A rollout restart of the Nginx controller Daemonset made the Ingresses work again instantly. Then it will work for a random period of time between 30 minutes and 3 hours and then it will fail again as all of you described above.

Only permanent fix is to set the ExternalTrafficPolicy back to cluster.

So basically my setup stopped working after migrating to containerd (where I don't see any possible relation since docker service is removed from nodes) and updating Kubernetes from 1.22.5 to 1.22.8 using Kubespray. So there's a good chance that some minor change in K8s could have brought this issue back.

To be honest it doesn't really make sense to me to have ExternalTrafficPolicy set to local and then access all ingresses via a MetalLB provided VIP because this VIP should be randomly hitting some node with a low chance to find the right container running there. But I didn't really care since it was working as per requirements. The traffic policy was set to Local and we saw SourceIPs in Logs. This was validated and went through QA controls flawlessly and suddenly it stopped working.

I think my best chance now is to further update K8s and see if it helps.

I somehow think it might be related to the Kubespray configured network stack based in calico. I saw strange Logs in NginX complaining about a non whitelisted VIP for the ingress and that it refuses to bind. But this might also be normal log noise since I never made a real log event baseline for my running stack.
